src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java

```diff
@@ -171,7 +200,70 @@ public void handleException(TransportException exp) {
+        } catch (Throwable e) {
+            logger.warn("failed to serialize cluster_state before publishing it to node {}", e, node);
+            publishResponseHandler.onFailure(node, e);
+            return;
+        }
+        try {
+            TransportRequestOptions options = TransportRequestOptions.options().withType(TransportRequestOptions.Type.STATE).withCompress(false);
+            // no need to put a timeout on the options here, because we want the response to eventually be received
+            // and not log an error if it arrives after the timeout
+            transportService.sendRequest(node, ACTION_NAME,
+                    new BytesTransportRequest(bytes, node.version()),
+                    options, // no need to compress, we already compressed the bytes
+
+                    new EmptyTransportResponseHandler(ThreadPool.Names.SAME) {
+
+                        @Override
+                        public void handleResponse(TransportResponse.Empty response) {
+                            if (timedOutWaitingForNodes.get()) {
+                                logger.debug("node {} responded for cluster state [{}] (took longer than [{}])", node, clusterState.version(), publishTimeout);
+                            }
+                            publishResponseHandler.onResponse(node);
+                        }
+
+                        @Override
+                        public void handleException(TransportException exp) {
+                            logger.debug("failed to send cluster state to {}", exp, node);
+                            publishResponseHandler.onFailure(node, exp);
+                        }
+                    });
+        } catch (Throwable t) {
+            logger.debug("error sending cluster state to {}", t, node);
+            publishResponseHandler.onFailure(node, t);
+        }
+    }
+
+    private BytesReference serializeFullClusterState(ClusterState clusterState, Version nodeVersion) throws IOException {
+        BytesStreamOutput bStream = new BytesStreamOutput();
+        StreamOutput stream = CompressorFactory.defaultCompressor().streamOutput(bStream);
+        stream.setVersion(nodeVersion);
+        stream.writeBoolean(true);
+        ClusterState.Builder.writeTo(clusterState, stream);
+        stream.close();
+        return bStream.bytes();
+    }
+
+    private BytesReference serializeDiffClusterState(Diff diff, Version nodeVersion) throws IOException {
+        BytesStreamOutput bStream = new BytesStreamOutput();
+        StreamOutput stream = CompressorFactory.defaultCompressor().streamOutput(bStream);
+        stream.setVersion(nodeVersion);
+        stream.writeBoolean(false);
+        diff.writeTo(stream);
+        stream.close();
+        return bStream.bytes();
+    }
+
     private class PublishClusterStateRequestHandler extends BaseTransportRequestHandler<BytesTransportRequest> {
+        private volatile ClusterState clusterState;
```

Comment:
can we call this last seen clusterState? it doesn't need to be volatile as it is changed under a lock.